ntoskrnl/io/pnpmgr/pnpirp.c

```diff
@@ -306,3 +306,62 @@ PiIrpQueryPnPDeviceState(
 
     return status;
 }
+
+// IRP_MN_QUERY_CAPABILITIES (0x09)
+NTSTATUS
+PiIrpQueryPnPDeviceCapabilities(
+    _In_ PDEVICE_NODE DeviceNode,
+    _Out_ PDEVICE_CAPABILITIES DeviceCaps)
+{
+    PVOID Dummy;
+    IO_STACK_LOCATION Stack = {
+        .MajorFunction = IRP_MJ_PNP,
+        .MinorFunction = IRP_MN_QUERY_CAPABILITIES
+    };
+
+    PAGED_CODE();
+
+    ASSERT(DeviceNode);
+
+    RtlZeroMemory(DeviceCaps, sizeof(DEVICE_CAPABILITIES));
+    DeviceCaps->Size = sizeof(DEVICE_CAPABILITIES);
+    DeviceCaps->Version = 1;
+    DeviceCaps->Address = -1;
+    DeviceCaps->UINumber = -1;
+    Stack.Parameters.DeviceCapabilities.Capabilities = DeviceCaps;
+
+    return IopSynchronousCall(DeviceNode->PhysicalDeviceObject, &Stack, &Dummy);
+}
+
+// IRP_MN_QUERY_ID (0x13)
+NTSTATUS
+PiIrpQueryPnPDeviceId(
+    _In_ PDEVICE_NODE DeviceNode,
+    _In_ BUS_QUERY_ID_TYPE IdType,
+    _Out_ PWCHAR *Id)
+{
+    NTSTATUS Status;
+    ULONG_PTR LongId;
```

Comment:
Why using that intermediate variable? Why not just passing the `Id` parameter directly? (as in the suggestion below).